noble-migration/src/bin/check.rs

```diff
@@ -0,0 +1,313 @@
+    }
+
+    // vec indexing is safe because we did the bounds check above
+    let free_space = parts[3]
+        .parse::<u64>()
+        .context("parsing free space failed")?;
+
+    Ok(free_space)
+}
+
+const EXPECTED_DOMAINS: [&str; 5] = [
+    // Ubuntu domains
+    "archive.ubuntu.com",
+    "security.ubuntu.com",
+    // SecureDrop domains (including testing ones)
+    "apt.freedom.press",
+    "apt-qa.freedom.press",
+    "apt-test.freedom.press",
+];
+
+/// Verify only expected sources are configured for apt
+fn check_apt() -> Result<bool> {
+    let output = process::Command::new("apt-get")
+        .arg("indextargets")
+        .output()
+        .context("spawning apt-get indextargets failed")?;
+    if !output.status.success() {
+        bail!(
+            "running apt-get indextargets failed: {}",
+            String::from_utf8_lossy(&output.stderr)
+        );
+    }
+
+    let stdout = String::from_utf8(output.stdout)
+        .context("apt-get stdout is not utf-8")?;
+    for line in stdout.lines() {
+        if line.starts_with("URI:") {
+            let uri = line.strip_prefix("URI: ").unwrap();
+            let parsed = Url::parse(uri)?;
+            if let Some(Host::Domain(domain)) = parsed.host() {
+                if !EXPECTED_DOMAINS.contains(&domain) {
+                    println!("apt: unexpected source: {domain}");
+                    return Ok(false);
+                }
+            } else {
+                println!("apt: unexpected source: {uri}");
+                return Ok(false);
+            }
+        }
+    }
+
+    println!("apt: all sources are expected");
+    Ok(true)
+}
+
+/// Check that systemd has no failed units
+fn check_systemd() -> Result<bool> {
+    let output = process::Command::new("systemctl")
+        .arg("is-failed")
+        .output()
+        .context("spawning systemctl failed")?;
+    if output.status.success() {
+        // success means some units are failed
+        println!("systemd: some units are failed");
+        Ok(false)
+    } else {
+        println!("systemd: no failed units");
+        Ok(true)
+    }
+}
+
+fn run() -> Result<()> {
+    let codename = os_codename()?;
+    if codename != "focal" {
+        println!("Unsupported Ubuntu version: {codename}");
+        // nothing to do, write an empty JSON blob
+        fs::write(STATE_PATH, "{}")?;
+        return Ok(());
+    }
+
+    let state = State {
+        ssh: check_ssh_group()?,
+        ufw: check_ufw_removed(),
+        free_space: check_free_space()?,
+        apt: check_apt()?,
+        systemd: check_systemd()?,
+    };
+
+    fs::write(
+        STATE_PATH,
+        serde_json::to_string(&state).context("serializing state failed")?,
+    )
+    .context("writing state file failed")?;
+    if state.is_ready() {
+        println!("All ready for migration!");
+    } else {
+        println!(
+            "Some errors were found that will block migration.
```

Comment:
Absolutely.

Reply:
I will spend some time working on the documentation and agree that we should link it here, but don't think it should hold up review/approval of this since it'll be trivial to add in afterwards but before the release.